Add conversion from v0.5.x to v0.6.0

b-reyes opened the pull request:
This PR allows users to open and use v0.5.x files in v0.6.0 of echopype. This was accomplished by adding a routine that automatically converts v0.5.x files into the format required for v0.6.0 (when `open_converted` is used). 


Note: this is currently under development.

emiliom commented on echopype/tests/preprocess/test_preprocess.py:
```diff
@@ -1,3 +1,5 @@
+import sys
```
Here and in `test_plot.py`, I assume you're going to use the `sys` package in work you'll be adding? Otherwise, it'd not currently being used.

emiliom commented on echopype/convert/set_groups_azfp.py:
```diff
@@ -167,6 +167,22 @@ def set_beam(self) -> xr.Dataset:
         ping_time = self.parser_obj.ping_time
         channel_id = self._create_unique_channel_name(unpacked_data)
 
+        # creates unique channel name
+        if unpacked_data["serial_number"].size == 1:
+            freq_as_str = unpacked_data["frequency"].astype(int).astype(str)
+
+            # TODO: replace str(i+1) with Frequency Number from XML
+            channel_id = [
+                str(unpacked_data["serial_number"]) + "-" + freq_as_str[i] + "-" + str(i + 1)
+                for i in range(len(freq_as_str))
+            ]
+        else:
+            channel_id = None
+            raise NotImplementedError(
+                "Creating a channel name for more than"
+                + " one serial number has not been implemented."
+            )
+
```
Is this block related to the 0.5x to 0.6x conversion functionality? It doesn't seem to be, as far as I can tell. 

I'm not suggesting there's anything inherently wrong with it. I just want to be clear about its purpose. It seems like it's an improvement unrelated to the core of this PR. Correct?

leewujung replied:
This block and the next one seem to be relics from sometime before? since the content is identical with what's in `_create_unique_channel_name`

b-reyes replied:
Yes, this should not be here. Thank you for catching this! It was caused by a merging of branches. I will go ahead and remove it.

emiliom commented:
In `test_echodata_structure.py` you use pre-generated converted files:

https://github.com/OSOceanAcoustics/echopype/blob/d95267d03a15da7d60f0e560baa577d0fcc378cc/echopype/tests/echodata/test_echodata_structure.py#L96-L102

But those files are not on the `test_data` image yet (I ran `setup-services.py --deploy` ~ 2pm). Could you upload them to Google Drive soon and manually run the GH action that updates the docker images?

b-reyes commented:
@leewujung and @emiliom I think this PR is in a state that you can review a good portion of the code. I have tried to document in the code all places where I may add items (these are marked with TODO). All files required for the conversion tests have been uploaded to the Google Drive. Please note that I will still need to add items once #681 and #682 have been resolved. For this reason, I will keep this in the draft state. 

Note, some tests are failing. I believe this is due to the fact that all tests are being ran with a `dev` version (as @emiliom remarked above). I will look further into this.

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -245,7 +244,9 @@ def _add_beam_dim(
                     .copy()
                 )
-                ds[var_name].beam.attrs = self._varattrs["beam_coord_default"]["beam"]
+                ds[var_name].beam.attrs = sonarnetcdf_1.yaml_dict["variable_and_varattributes"][
+                    "beam_coord_default"
+                ]["beam"]
```
With this change it seems line 47 `self._varattrs = sonarnetcdf_1.yaml_dict["variable_and_varattributes"]` should be removed. It is not used anywhere in the code.

b-reyes replied:
Line 47 should not be removed. `self._varattrs` is used in the specific `set_groups_X.py` for the sensors. It is necessary to use this line because this method should be static. We can't use `self._varattrs` here because this function is used in my conversion routines, which do not initialize this group (thus `self._varattrs` does not exist).

leewujung replied:
Ah I see, thanks for the explanation!

leewujung commented on echopype/echodata/sensor_ep_version_mapping/ep_version_mapper.py:
```diff
@@ -0,0 +1,30 @@
+from .v05x_to_v06x import convert_v05x_to_v06x
+
+
+def map_ep_version(echodata_obj):
```
I like how you have made this flexible for expansion later!  :D
(Not that we want a lot of these...)

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1091 @@
+            # rename frequency to channel
+            ed_obj[grp_path]["frequency_nominal"] = ed_obj[grp_path].frequency
+
+            if "standard_name" in ed_obj[grp_path].frequency_nominal.attrs:
+                del ed_obj[grp_path].frequency_nominal.attrs["standard_name"]
```
I think we can keep this, since the nominal frequency _is_ sound frequency.

emiliom replied:
Yes, keep
```suggestion
```

b-reyes replied:
This attribute was dropped when we created this variable. So, we need to do a PR that adds this attribute back in (occurs in `set_groups_X.py`). If we all agree that this should be included, is the following attribute appropriate (this is what it was): 

`frequency_nominal.attrs["standard_name"] = "sound_frequency"`

leewujung replied:
Yes, please go with that. You can self-merge that PR to speed this up.

emiliom replied:
yes

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1091 @@
+
+def get_channel_id(ed_obj, sensor):
+    """
+    Returns the channel_id for all non-unique frequencies.
```
Could you elaborate on what you meant by "non-unique" here? Do you mean just all the `channel_id` will be included and concatenated?

b-reyes replied:
This function will grab all channel ids, even those channel ids where the frequencies are repeated (that's why I say "non-unique frequencies"). If you have a better sentence to describe this, please let me know.

emiliom replied:
Going back to a comment found elsewhere, there will be no 0.5.x converted files with repeated frequencies. 0.5.x could not generate such files.  So, there's no need to test or account for that case.

b-reyes replied:
To clarify a little further, I specifically state this for cases such as the EK80 sensor. When we have two beam groups, we may have frequencies that are the same, but have a different channel id. Note that I am collecting the channel ids from all beam groups (for EK80).

emiliom replied:
I see. I had not considered that case

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1091 @@
+
+    else:
+
+        channel_id = xr.concat(
+            [child.ds.channel_id for child in ed_obj._tree["Sonar"].children], dim="frequency"
```
I haven't tried to evaluate this myself, but since dim is set to `frequency` and `xr.concat` is used directly, I am curious if this works in the case with duplicated frequencies. BUT, we won't have a v0.5.x file with such characteristics because it won't convert, so this is moot.

emiliom replied:
> BUT, we won't have a v0.5.x file with such characteristics because it won't convert, **so this is moot.**

Indeed

b-reyes replied:
I am unsure what will happen with repeated frequencies. As to not spend time on this, I will not look into this right now (since we will not encounter this scenario).

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1091 @@
+    3. Remove the variables ``beamwidth_receive_athwartship``
+    and ``beamwidth_transmit_alongship``
+    4. Change the attribute ``long_name`` in the variables
+    ``angle_offsetalongship/athwartship`` and
```
```suggestion
    ``angle_offset_alongship/athwartship`` and
```

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1091 @@
+            ["beamwidth_receive_athwartship", "beamwidth_transmit_alongship"]
+        )
+
+        ed_obj["Sonar/Beam_group1"].angle_sensitivity_alongship.attrs[
+            "long_name"
+        ] = "alongship angle sensitivity of the transducer"
+
+        ed_obj["Sonar/Beam_group1"].angle_sensitivity_athwartship.attrs[
+            "long_name"
+        ] = "athwartship angle sensitivity of the transducer"
```
Small things, but for EK80 in v0.5.x both of these miss one word "angle" in it, so it seems a good idea to just reset:
https://github.com/OSOceanAcoustics/echopype/blob/bc8afa190fa2ca4fab5944bac83cd4b20f7abcf6/echopype/convert/set_groups_ek80.py#L300-L309

b-reyes replied:
So, you are suggesting to do a PR that removes "angle" from EK60. Or we could just add "angle" to EK80.

leewujung replied:
I am thinking to add "angle" to EK80.

b-reyes replied:
Ok, I will go ahead and do this.

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1091 @@
+                ] = beam_group.ds.angle_offset_athwartship.attrs["comment"]
+
+
+def _beam_groups_to_convention(ed_obj, set_grp_cls):
```
May be a good idea to make this `_beamgroups_to_convention` to be consistent with what you have in `set_grp_cls.beamgroups_to_convention`

b-reyes replied:
Are you suggesting that I rename this function to `_beamgroups_to_convention`?

leewujung replied:
I just think it should be consistent in all places. My inclination is `_beam_groups_*` because the groups are called `Beam_groupX`.

b-reyes replied:
Ok, I will change `set_grp_cls.beamgroups_to_convention` to `set_grp_cls.beam_groups_to_convention`.

emiliom commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1131 @@
+
+        if "frequency" in ed_obj[grp_path]:
+
+            # add frequency_nominal, remove standard_name attribute, and
+            # rename frequency to channel
```
Adjusting comment to reflect the suggested change in line 166
```suggestion
            # add frequency_nominal and rename frequency to channel
```

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1131 @@
+
+    ed_obj["Platform"] = xr.merge([ed_obj["Platform"], ds_tmp])
+
+    if sensor != "AZFP":
```
Just adding comment for our later selves:
```suggestion
    if sensor != "AZFP":  # this variable was missing for AZFP v0.5.x
```

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1131 @@
+                full_transducer_vars[spatial], dim="channel"
+            )
+
+    if sensor == "EK80":
+        ed_obj["Platform"]["frequency_nominal"] = ed_obj["Vendor"].frequency_nominal.sel(
+            channel=ed_obj["Platform"].channel
+        )
```
My previous comment might have been washed out, but I am thinking that now we have everything aligned with `channel`, do we still need `frequency_nominal` to be in this dataset? I can see this makes EK80 consistent with EK60 though, maybe that's the rationale here?
I am fine with keeping it here since it has a small footprint.

b-reyes replied:
In previous issues, we explicitly state that we should add `frequency_nominal` to every dataset that has `channel`. I think this is a good idea because then you don't have to go looking for what frequency corresponds to the channel. Furthermore, I think `Platform.frequency_nominal` is referenced in downstream processes. I highly suggest that we do not remove this.

emiliom replied:
I agree with @b-reyes . If `channel` is present, so should `frequency_nominal`

leewujung replied:
Sounds good.

leewujung commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1131 @@
+
+    for grp_path in ed_obj.group_paths:
+
+        if "location_time" in list(ed_obj[grp_path].coords):
+
+            # renames location_time to time1
+            ed_obj[grp_path] = ed_obj[grp_path].rename(name_dict={"location_time": "time1"})
+
+        if "mru_time" in list(ed_obj[grp_path].coords):
+
+            # renames mru_time to time2
+            ed_obj[grp_path] = ed_obj[grp_path].rename(name_dict={"mru_time": "time2"})
```
Seems that we need to attach these time coordinates with the nice set of v0.6.0 attributes also?

leewujung replied:
I see that you have another 2 functions `_add_time_comment_in_environment` and `_add_time_comment_in_platform` for this, but if I am not mistaken, those only deal the "comment" attribute. From #682 it seems `long_name` needs update too, so may be a good idea to just copy the entire new set, perhaps in the functions that you already have?

b-reyes replied:
We have not arrived at a consensus for #682 yet. Once we have, I will implement it. I planned on putting the changes in the functions you mention and renaming them (I haven't fully thought this out though).

leewujung reviewed (commented):
@b-reyes : Thanks for this good chunk of work! I was so happy when getting to the last function 🙃! 

I've put inline comments and suggestions, but feel free to not use the direct commit suggestions if that makes it easier for you to keep track of changes. GH really needs to improve this!

emiliom commented:
It looks like your code is only accounting for AZFP, EK60 and EK80 (and AD2CP, in order to explicitly ignore it) sensor strings, strictly speaking. It'll miss the less common and "tricky" sensor models "ES70", "ES80", "EA640". What makes these tricky is that these sensors reuse the parser and setgroup from EK60 and EK80, rather than having their own. This relationship is specified in [core.SONAR_MODELS](https://github.com/OSOceanAcoustics/echopype/blob/dev/echopype/core.py#L41). I suggest that tests such as this one:
https://github.com/OSOceanAcoustics/echopype/blob/c8ce04055b1234035c9e245d000bbc3ad6163246/echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py#L217
be changed to tests based on `SONAR_MODELS[sensor]["set_groups"]` instead. It's probably cleaner if you set up small utility function to do the testing for you. Conceptually, what you'll be testing for is the corresponding SetGroupsX class.

Related to this, I'm not 100% sure, but I think echopype uses `sensor_model` fairly consistently to represent the specific sonar model strings (eg, ES70) as listed in [SonarModelsHints](https://github.com/OSOceanAcoustics/echopype/blob/dev/echopype/core.py#L19). To make this meaning more explicit, I suggest you rename the variable `sensor` in `v05x_to_v06x` to `sensor_model`. Initially, I wasn't entirely sure if `sensor` already accounted for the mapping from `sensor_model` to its corresponding SetGroupsX (eg, ES70 > EK60).

emiliom commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1131 @@
+        if "frequency" in list(ed_obj[grp_path].coords):
+
+            # creates consistent frequency attributes
+            # NOTE: the attribute standard_name will be removed in `frequency_nominal`
```
Delete this comment line. See suggestion elsewhere about retaining standard_name
```suggestion
```

emiliom commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1131 @@
+
+def _reorganize_beam_groups(ed_obj):
+    """
+    Maps Beam --> Sonar/Beam_group1 and Beam_power --> Sonar/Beam_group2.
```
I'm not 100% sure about my recollection (sigh), but I believe an EK80 can have power data only. In that case, I think the only beam group that would exist is Beam_power, and it would map to Sonar/Beam_group1 rather than Sonar/Beam_group2. @leewujung can you confirm?

b-reyes replied:
@emiliom I think you are referring to a case I highlighted in your previous [PR](https://github.com/OSOceanAcoustics/echopype/pull/658#discussion_r867061536). If this was what you are indeed referring to, it is not a problem. In v0.5.6 we set this value to `ed.beam`, so this case would get mapped to `Sonar/Beam_group1` (this is correct, it should not be `Sonar/Beam_group2`). This is what should happen we accounted for this in your PR by adding "(either complex samples or uncalibrated power samples)" to the description.

emiliom replied:
Great, thank you.

emiliom commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1131 @@
+
+            else:
+                ed_obj[grp_path]["channel"] = channel_id.sel(
+                    frequency=ed_obj[grp_path].frequency_nominal
```
Is this `sel` necessary?

b-reyes replied:
Yes, `sel` is necessary. In the case of EK80 I collect all of the channel ids (the `channel_id` variable above) and these frequencies may be out or order. This is a problem when we go to set `channel` for groups like `Vendor_specific`, which need all of the frequencies. However, groups like `Vendor_specific` have their `frequency` in ascending order, so if we did not include `sel` here, the "channels" can be potentially lined up with the wrong frequency.

leewujung replied:
I experienced this while working on #697.

emiliom commented on echopype/echodata/sensor_ep_version_mapping/v05x_to_v06x.py:
```diff
@@ -0,0 +1,1131 @@
+                del ed_obj[grp_path].frequency_nominal.attrs["standard_name"]
+            ed_obj[grp_path] = ed_obj[grp_path].rename({"frequency": "channel"})
+
+            # set values for channel
```
`ed_obj[grp_path]["channel"]` is float type, b/c it's simply the old `frequency` renamed. But here it's assigned string values. I'm either puzzled or pleasantly surprised that the assignment doesn't lead to an error. Is the `ed_obj[grp_path]["channel"]` data type being changed to string?

b-reyes replied:
Yes, I was very happy that xarray took these types of scenarios into consideration. If you are assigning a coordinate to a new array that is the same size, no error will occur even if they are of different types. I imagine that xarray points to positions in the coordinate array, rather than hard coding them in.

b-reyes commented:
> It looks like your code is only accounting for AZFP, EK60 and EK80 (and AD2CP, in order to explicitly ignore it) sensor strings, strictly speaking. It'll miss the less common and "tricky" sensor models "ES70", "ES80", "EA640".

@emiliom thank you for catching this. I incorrectly assumed that `keywords` contained the EchoData structure that was used. I have created a function that now maps the sensor model to the appropriate EchoData structure used.

emiliom commented:
@b-reyes can you tell us which 0.5 > 0.6 conversion tests are working vs which ones are not, at this time? No need to elaborate. 

I'm finding it nearly impossible, realistically, to verify that every single conversion step is what we settled on. There is a ton! And it may not be fully necessary if the tests of a 0.5 > 0.6 conversion vs the direct raw > 0.6 converted file are doing their job. Thanks.

b-reyes commented:
> @b-reyes can you tell us which 0.5 > 0.6 conversion tests are working vs which ones are not, at this time? No need to elaborate.
> 
> I'm finding it nearly impossible, realistically, to verify that every single conversion step is what we settled on. There is a ton! And it may not be necessary to do that if the tests of a 0.5 > 0.6 conversion vs the direct raw > 0.6 converted file are doing their job. Thanks.

Right now `test_v05x_v06x_conversion_structure` accounts for all changes to `dev` so far. More specifically, at this point everything besides #682 has been taken into account and tested for in `test_v05x_v06x_conversion_structure`. Be sure to download the newest Google Drive files as I changed them this morning.

emiliom commented on echopype/tests/echodata/test_echodata_structure.py:
```diff
@@ -0,0 +1,271 @@
+    """
+
+    converted_raw_paths_v06x = [ek60_path / "ek60-Summer2017-D20170615-T190214-ep-v06x.nc",
+                                ek80_path / "ek80-Summer2018--D20180905-T033113-ep-v06x.nc",
```
We probably should identify a couple of contrasting EK80 raw files to test, rather than just one, to catch EK80 variability; eg, one having both complex and power data, vs just power.

leewujung replied:
I agree. They are explicitly listed in the tests in `tests/calibrate/test_calibrate.py`

b-reyes replied:
I have added extra EK80 raw files to test in my most recent push

emiliom commented:
I thought this review would never end :exploding_head: ! Wow, this is a lot of stuff. Great job methodically keeping track of everything! You laid out a nice structure that, as everyone has said, hopefully we'll never have to reuse.

Now it'll just be a matter of seeing the tests working. I didn't go through test_echodata_structure.py in much detail. Once the tests are working, I'll pay closer attention.

b-reyes commented:
@emiliom and @leewujung in my most recent push I added several test files for the conversion. Some of these include EK80 files (agreed upon by @leewujung and myself) and also a combined file. While adding these test files I found a couple of small issues with my conversion. I have corrected these small issues, which also required mild changes to the conversion tests. All of these files have been added to the Google Drive and I have reran the Git Action for this. 

Additionally, while trying out these test files I found an issue in `set_groups_ek80.py`, specifically when we set the `Platform` group (I will put a comment on this change). Essentially, before we were collecting all of the channel ids in this group, I found that we should not do this because in some cases we only have the power data. I have corrected this so that it will account for all of these cases. 

Note, the tests are still failing in this PR. As stated above, I believe this is due to the fact that all tests are being ran with a `dev` version (as @emiliom remarked above). I am not sure the best way to deal with this, but @leewujung said she would look into it too.

This PR is ready for either a final review or approval.

b-reyes commented on echopype/convert/set_groups_ek80.py:
```diff
@@ -210,7 +210,8 @@ def set_sonar(self, beam_group_count=1) -> xr.Dataset:
         """Set the Platform group."""
 
-        ch_ids = list(self.parser_obj.config_datagram["configuration"].keys())
+        ch_ids = self.parser_obj.ch_ids["complex"] + self.parser_obj.ch_ids["power"]
```
Note: before we were collecting all of the channel ids in this group, I found that we should not do this because in some cases we only have the power data. I have corrected this so that it will account for all of these cases.

emiliom replied:
I'm not familiar with the parser, so I don't know what would be in `self.parser_obj.config_datagram["configuration"].keys()` compared to `self.parser_obj.ch_ids["complex"] + self.parser_obj.ch_ids["power"]`. But I ran `open_raw` on a power-only EK80 raw file using the version of echopype currently in `dev` vs what's in your PR. The results for `channel` and `frequency_nominal` were identical. I then ran that comparison on an EK80 file that has both complex and power data. In that case, the only difference was one I assume is not important: `channel` and `frequency_nominal` were sorted differently, but the two still matched up the same way.

So, while I don't understand the problem you're trying to solve, your change doesn't seem to introduce any problems. I'm good with merging.

emiliom replied:
I'll just add that I looked at `frequency_nominal` mainly for reference and b/c it was in the next code block.

b-reyes replied:
Basically all of the channels occur in `self.parser_obj.config_datagram["configuration"].keys()`, and for `self.parser_obj.ch_ids["complex"] + self.parser_obj.ch_ids["power"]` we only obtain the channels that are in the beam groups. In one of the files I included in this PR (to test the conversion), I found a case where these two statements are not equal. Hopefully this is just a corner case, but either way we now take it into account.

> In that case, the only difference was one I assume is not important: channel and frequency_nominal were sorted differently in dev vs your PR, but the two still matched up the same way (ie, frequency values matched up with the same channel).

Since we go on to use the channel id to select the frequency and the other variables, the order does not matter (as you assumed).

emiliom commented:
I scanned through your commits since yesterday. In general, they look good. But in some areas I'm not in a position to assess it (channel ids and related changes); for the tests, it's hard to sink my teeth into them w/o seeing them in action.

> Note, the tests are still failing in this PR. As stated above, I believe this is due to the fact that all tests are being ran with a dev version (as @emiliom remarked above). I am not sure the best way to deal with this, but @leewujung said she would look into it too.

The `echodata.EchoData.version_info` code suggestion (hack) I added in  https://github.com/OSOceanAcoustics/echopype/pull/685#issuecomment-1130742396 should do the trick, but that means you'd have to create all your 0.6.0 nc files again, with this hack. I thought you said you had manually set `Provenance.conversion_software_version` to "0.6.0"? Anyway, whatever is easiest for you. IMHO my hack could go into the release, and we can remove it from `dev` immediately after the release. It has no negative consequence until a 0.6.1 is released.

b-reyes commented on echopype/utils/prov.py:
```diff
@@ -17,9 +18,10 @@ def echopype_prov_attrs(process_type: ProcessType) -> Dict[str, str]:
         Echopype process function type
     """
+    # TODO: change hard coded 0.6.0 after release
     prov_dict = {
         f"{process_type}_software_name": "echopype",
-        f"{process_type}_software_version": ECHOPYPE_VERSION,
+        f"{process_type}_software_version": "0.6.0",  # ECHOPYPE_VERSION,
```
We need to change this after the release. It is hard coded to 0.6.0 right now so that tests pass.

b-reyes replied:
@emiliom if you don't mind, I would like to leave this as unresolved so that we can refer to it in the future. Is that ok?

b-reyes commented:
The reason these test were failing were for two reasons:
1. We have a test that opens a file, saves it to a netcdf, and then opens this netcdf. The issue with the dev version is that it will have version 0.5.6 right now. So, when we go to open it, my converter kicks in and this causes the tests to fail. The quick fix is to set the software version to 0.6.0 and then remove it after the release (I have done this and I point it out above). 
2. In one of the tests we have a file `Summer2017-D20170615-T190214__NEW.zarr` that needed to be updated to v0.6.0. I have updated this file and put it in the Google drive. I renamed the previous file to `Summer2017-D20170615-T190214__NEW_old.zarr` (I know this isn't a great name).

Hopefully all of the tests will pass now (they are passing locally).

emiliom commented on echopype/tests/echodata/test_echodata_structure.py:
```diff
@@ -0,0 +1,305 @@
+        print(f"path_v05x = {path_v05x}")
+
+        ed_v05x = open_converted(path_v05x)
+        tree_v06x = _tree_from_file(converted_raw_path=path_v06x)
```
What's the reason for opening the 0.6.x files via datatree rather than via `open_converted`? As far as I can see, the results should be the same. But it looks like using `open_converted` would make `_tree_from_file` unnecessary and would make for a slightly better test (ie, actually opening a 0.6.x file via echopype's `open_converted` machinery).

b-reyes replied:
The reason I did not use `open_converted` to open the 0.6.x file is because I wanted the Datatree created to be as independent from the `EchoData` object as possible. I do this just incase we happen to make modifications to `EchoData` in the future that would somehow change the tree structure. I am not sure if this is absolutely necessary, but I do not see the harm in having it either.

emiliom replied:
Thanks. Certainly no harm.

b-reyes commented on echopype/tests/echodata/test_echodata_structure.py:
```diff
@@ -0,0 +1,305 @@
+            for var in val:
+                _check_and_drop_attr(ed_v05x, tree_v06x, key, var[0], var[1])
+
+        _check_and_drop_var(ed_v05x, tree_v06x, "Provenance", "source_filenames")
```
The reason these tests are passing is because this global attribute is deleted in `_add_source_filenames_var` in the conversion routine.

emiliom replied:
Of course, I realize that now that you say it! My apologies for the dumb comment. I guess seeing the statement
`ed_v05x["Provenance"] = ed_v05x["Provenance"].drop("src_filenames")` for the combined file sent me thinking in the wrong direction.

b-reyes replied:
No worries!

emiliom commented:
I'm done reviewing the areas I'm in a position to review; that excludes changes related to channel id's. With more time, tomorrow, I may be able to wrap my head around it. But maybe @leewujung can tackle those changes tomorrow.

Otherwise, it's looking great and it's almost ready!

emiliom commented:
@b-reyes I've approved this PR. I suggest we merge. If anything comes up, we can always address it directly rather than holding up the merging of this PR.
